packages/sdk-rtl/src/paging.ts

```diff
@@ -0,0 +1,405 @@
+        name: arrRes[2].trim(),
+      }
+    }
+  })
+  return obj
+}
+
+/** Event to observe the paging call */
+export type PageObserver<TSuccess> = (
+  /** Current retrieved page of results */
+  page: TSuccess
+) => TSuccess
+
+/**
+ * Create an API response pager for an endpoint that returns a Link header
+ * @param sdk implementation of IAPIMethods. Can be full SDK or functional auth session
+ * @param func sdk call that includes a paging header
+ * @param options transport options override to capture and use in paging requests
+ *
+ * @remarks `TSuccess` must be a collection type that supports `length`
+ */
+export async function pager<TSuccess extends ILength, TError>(
+  sdk: IAPIMethods,
+  func: PagingFunc<TSuccess, TError>,
+  options?: Partial<ITransportSettings>
+): Promise<IPager<TSuccess, TError>> {
+  return await new Paging<TSuccess, TError>(sdk, func, options).init()
+}
+
+/**
+ * Create an API response pager and collect all pages, returning the result
+ * @param sdk implementation of IAPIMethods. Can be full SDK or functional auth session
+ * @param func sdk call that includes a paging header
+ * @param onPage observer of the latest page of results. Defaults to noop.
+ * @param options transport options override to capture and use in paging requests
+ */
+export async function pageAll<TSuccess extends ILength, TError>(
+  sdk: IAPIMethods,
+  func: PagingFunc<TSuccess, TError>,
+  onPage: PageObserver<TSuccess> = (page: TSuccess) => page,
+  options?: Partial<ITransportSettings>
+): Promise<SDKResponse<TSuccess, TError>> {
+  const paged = await pager(sdk, func, options)
+  let rows: any[] = []
+  rows = rows.concat(onPage(paged.items))
+  let error
+  try {
+    while (paged.more()) {
+      const items = await sdk.ok(paged.nextPage())
+      rows = rows.concat(onPage(items))
+    }
+  } catch (err) {
+    error = err
+  }
+  if (error) {
+    return { ok: false, error }
+  }
+  return { ok: true, value: rows as unknown as TSuccess }
+}
+
+/**
+ * Link header pages class
+ */
+export class Paging<TSuccess extends ILength, TError>
+  implements IPager<TSuccess, TError>
+{
+  items: TSuccess = [] as unknown as TSuccess
+  links: PageLinks = {}
+  total = -1
+  offset = -1
+  limit = -1
+
+  private transport: BaseTransport
+
+  /**
+   * Create an API paginator
+   * @param sdk functional AuthSession or full SDK implementation
+   * @param func sdk function to call
+   * @param options transport overrides to use for subsequent requests
+   */
+  constructor(
+    public sdk: IAPIMethods,
+    public func: PagingFunc<TSuccess, TError>,
+    public options?: Partial<ITransportSettings>
+  ) {
+    this.transport = sdk.authSession.transport as BaseTransport
+  }
+
+  private async rawCatch(func: () => any) {
+    let raw: IRawResponse = {} as IRawResponse
+    const saved = this.transport.observer
+    try {
+      this.transport.observer = (response: IRawResponse) => {
+        raw = response
+        return response
+      }
```

Comment:
The TypeScript pattern I've found (and used from before TypeScript) is to do resource assignments that will be closed/restored inside the `try` section, which IMO helps clarify that `finally` is intended to restore or close them. Happy to discuss further.

The only thing happening is the capture of the observer from whatever transport is used for resolving the initial SDK call (functional or monolithic). I think we need to capture the SDK function's transport and intercept the call rather than creating a *different* transport that may be completely missing some of the configuration of the SDK's transport object.

Reply:
> to do resource assignments that will be closed/restored inside the `try` section

then I'd argue that perhaps `const saved = this.transport.observer` belongs inside the `try`?

> resolving the initial SDK call

ah, right - there's this asymmetry between the implementation of the initial call and subsequent `nextPage()`-and-friends calls. I feel like I want to think about that a little more.

possibly related is the asymmetry you see in my example code: I need to have an initial "get the items" block of code before I can launch into a while loop of retrieving the remaining items from subsequent pages if there are any. Maybe there's a smoother way to write that example? or maybe that will come with the "row-at-a-time"/generic iterator? just thinking out loud for our in-person chat on this

Reply:
The problem with having to capture the first set of items outside the loop  is that we won't have any `Link` header until we make the initiating request that causes paging. However, this means that we will *also* be sure to get the requested rows for any SDK method that does *not* include the paging headers. Having to init a stateful object for subsequent paging requests is a necessary evil.

Maybe there's a cleaner way to do it, but I haven't figured one out yet.

Reminds me we should make sure `pageAll` works with a non-paging sdk request, too. I'll add a functional test for that right now.